python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -162,8 +161,17 @@ def read_fn():
         self.last_update_time = 0.0
         self.update_interval_s = update_interval_s
 
-        # Tracks active worker nodes
+        # The following block of data on active nodes is filled once per
+        # autoscaler update in `self.update_node_lists()`.
+        # Managed worker nodes (node kind "worker"):
         self.workers = []
+        # Managed and unmanaged worker nodes
+        # (node kinds "worker" and "unmanaged"):
+        self.all_workers = []
+        # All nodes (workers, unmanaged workers, and head)
+        self.all_nodes = []
```

Comment:
Can you please add type annotation here?

Reply:
are these ips? Why we need all_nodes, shouldn't workers and all_nodes be enough?

Reply:
resource demand scheduler takes the list of all nodes as input
An equivalent way of doing it would be to get "head_node" and "all_workers" and then extract the ip of the head node using a node provider call.

Slightly more aesthetic, but requires making a couple extra provider.internal_ip calls which is a tiny bit of overhead for some node providers.

Will do it that way.

Reply:
There's a bad habit in the code, which I've absorbed, of calling node id variables "nodes", which makes it unclear what the variable is. Will change to "node ids".

Reply:
Annotations added.